cli/command/stack/kubernetes/services.go

```diff
@@ -2,43 +2,136 @@ package kubernetes
 
 import (
 	"fmt"
+	"sort"
+	"strings"
 
 	"github.com/docker/cli/cli/command/formatter"
 	"github.com/docker/cli/cli/command/stack/options"
 	"github.com/docker/cli/kubernetes/labels"
+	"github.com/docker/docker/api/types/filters"
+	appsv1beta2 "k8s.io/api/apps/v1beta2"
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+var supportedServicesFilters = map[string]bool{
+	"mode":  true,
+	"name":  true,
+	"label": true,
+}
+
+func generateValuedLabelsSelector(valuedLabels map[string][]string) []string {
+	var result []string
+	for k, v := range valuedLabels {
+		if len(v) == 1 {
+			result = append(result, fmt.Sprintf("%s=%s", k, v[0]))
+		} else {
+			sort.Strings(v)
+			result = append(result, fmt.Sprintf("%s in (%s)", k, strings.Join(v, ",")))
+		}
+	}
+	return result
+}
+
+func parseLabelFilters(rawFilters []string) ([]string, map[string][]string) {
+	var presentLabels []string
+	valuedLabels := map[string][]string{}
+	for _, rawLabel := range rawFilters {
+		equalIndex := strings.Index(rawLabel, "=")
```

Comment:
This can be simplified by using `strings.SplitN()`;


```go
func parseLabelFilters(rawFilters []string) ([]string, map[string][]string) {
	var presentLabels []string
	valuedLabels := map[string][]string{}
	for _, rawLabel := range rawFilters {
		v := strings.SplitN(rawLabel, "=", 2)
		key := v[0]
		if len(v) > 1 {
			valuedLabels[key] = append(valuedLabels[key], v[1])
			continue
		}
		presentLabels = append(presentLabels, key)
	}
	return presentLabels, valuedLabels
}
```

However, filters can appear both as `valuedLabel` and as `presentLabel`, is that expected? see https://play.golang.org/p/EuJk1d7b5Ts

If that's not the intent, we could treat `presentLabels` and `valuedLabels` as identical; a `presentLabel` just being a `valuedLabel` without value, and change `generateValuedLabelsSelector()` to `generateSelector()`, something like;

```go
func generateSelector(labels map[string][]string) []string {
	var result []string
	for k, v := range labels {
		switch len(v) {
		case 0:
			result = append(result, k)
		case 1:
			result = append(result, fmt.Sprintf("%s=%s", k, v[0]))
		default:
			sort.Strings(v)
			result = append(result, fmt.Sprintf("%s in (%s)", k, strings.Join(v, ",")))
		}
	}
	return result
}
```

Reply:
> However, filters can appear both as `valuedLabel` and as `presentLabel`, is that expected? see https://play.golang.org/p/EuJk1d7b5Ts

Good point, on `docker`/`swarm` it's basically a `and`, i.e. `--filter=label=foo --filter=label=foo=bar` is the same as `--filter=label=foo=bar`. I think we should do/keep the same, and thus `presentLabels` and `valuedLabels` shouldn't have anything in common :angel: 

So yeah, we should tread them as identical and update/fix `generateSelector` instead :+1: